src/lib/controllers/config-manifest/config-manifest.ts

```diff
@@ -0,0 +1,231 @@
+/*
+Copyright 2019 Balena Ltd.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+import * as Ajv from 'ajv';
+import * as _ from 'lodash';
+
+import { loadFromUri } from '../../tools';
+import {
+	ConfigManifestSchema,
+	ConfigManifestKeys,
+} from './config-manifest-schema';
+import { path } from 'temp';
+import { fs } from 'mz';
+import { type } from 'os';
+// tslint:disable-next-line:no-var-requires
+const configManifestSchemaJson = require('../../../../schemas/config-manifest-schema.json');
+
+const ajv = new Ajv();
+const configManifestSchema = ajv.compile(configManifestSchemaJson);
+
+/**
+ * ConfigManifest class
+ * Used for transforming/exposing a config-manifest
+ */
+// WIP notes: Kept this only because it provides the ability to return either
+// 	a JSON schema or valid ConfigManifest at this point
+export type NameValueCollection = { [propName: string]: any };
+
+export type ConfigManifestRaw = {
+	version: string;
+	title: string;
+	properties: ConfigManifestKeys;
+};
+
+export class InvalidManifestSchemaError extends Error {
+	public errors: Ajv.ErrorObject[];
+
+	constructor(errors: Ajv.ErrorObject[] | null = []) {
+		super('Invalid manifest schema');
+		this.errors = errors === null ? [] : errors;
+	}
+}
+
+export class ConfigManifest {
+	/**
+	 * Creates ConfigManifest using:
+	 * @param {string} productRepoURI
+	 * @param {string} path
+	 * @returns {Promise<ConfigManifest>}
+	 */
+	static async create(
+		productRepoURI: string,
+		paths: string[] = [],
+	): Promise<ConfigManifest> {
+		if (paths.length === 0) {
+			paths = ['config-manifest.yml'];
+		}
+
+		const schemas = await Promise.all(
+			paths.map(
+				p =>
+					loadFromUri({
+						uri: productRepoURI,
+						path: p,
+						errorMessage:
+							'Unable to find config-manifest.yml. See documentation_link for more info.\n',
+					}) as Promise<ConfigManifestRaw>,
+			),
+		);
+		return new ConfigManifest(
+			schemas.map(schema => {
+				return { schema };
+			}),
+		);
+	}
+
+	/**
+	 * Applies when condition to JsonSchema
+	 * @param obj
+	 * @param {string} key
+	 */
+	private static applyWhenCondition(obj: any, key: string): void {
+		const properties = _.get(obj, key);
+		const conditions: any = {};
```

Comment:
```suggestion
		const conditions: Dictionary<any> = {};
```

Reply:
which `Dictionary` type are you referring to here? Lodash has one, but it doesn't implement the iterator type for the below `for ... of ...` block.

Am I missing something?

Reply:
The standard `Dictionary` we define all over the place:
```ts
interface Dictionary<T> {
	[key: string]: T;
}
```

Reply:
@Page- makes sense, just checking I wasn't being daft. I'll define the `Dictionary<T>` type and check it's used instead of being redefined all the time (which I think it is now)